### ISSUE TITLE
Feature: ".tslintignore" file to exclude globs

### ISSUE THREAD

**Issue by am11** [palantir/tslint#73]
### Update (2016/06/01)

`--exclude` CLI flag is available since TSLint v3.6. This issue now tracks a feature request for a `.tslintignore` file.
### Original issue

Cc: https://github.com/madskristensen/WebEssentials2013/issues/617

From [JsHint docs](http://www.jshint.com/docs/cli/), there are two flags for exclusion in cli options:

> --exclude
> 
> Allows you to specify directories which you DON'T want to be linted.
> 
> --exclude-path
> 
> Allows you to provide your own .jshintignore file. For example, you can point JSHint to your .gitignore file and use it instead of default .jshintignore.

Please provide us the similar flags, so we can pass .tslintignore file to exclude files and folders when some batch / automated executor script runs lint on it.

See JsHint's implementation of this feature: [cli.js#L182-L208](https://github.com/jshint/jshint/blob/56a4a78dd65a872d93d843621c14d2ebe575cfed/src/cli.js#L182-L208) and its caller; [gather function](https://github.com/jshint/jshint/blob/56a4a78dd65a872d93d843621c14d2ebe575cfed/src/cli.js#L411-L435).

**Comment by pgrm** [comment]
any updates on this?

**Comment by JKillian** [comment]
This seems like a relatively important feature that ought to be prioritized. Wanted to note interest in this feature from a [SO question](http://stackoverflow.com/questions/34578677/how-to-ignore-a-particular-directory-or-file-for-tslint)

**Comment by adidahiya** [comment]
Might be better to support this via tsconfig.json (#858) rather than re-implementing this in a CLI flag...

**Comment by JKillian** [comment]
Hmm, the difficulty with that is that you may want to compile different files than you lint. So I could potentially see using `tsconfig.json` and then also using `--exclude` to not lint a few of the normally included files.

**Comment by bill-mybiz** [comment]
As opposed to CLI, would this not be a natural fit in the tslint.json file and then add an ignore check that parses the "ignorePaths" at https://github.com/palantir/tslint/blob/master/src/tslint.ts#L46?

**Comment by tsm91** [comment]
+1

**Comment by shaharmor** [comment]
+1, a .tslintignore will be great

**Comment by lukaselmer** [comment]
+1 for .tslintignore or add an option to ignore certain paths in tsconfig.json

It could be similar to https://github.com/bbatsov/rubocop#includingexcluding-files

**Comment by adidahiya** [comment]
As a first pass, we'll merge #1006 soon, which adds a simple CLI `--exclude` option. Does that get you most of the way there? If you're using additional tooling on top of `tslint`, then it probably has glob / exclude support (grunt / gulp / etc). How important is this to support in `tslint.json`?

**Comment by radek-holy** [comment]
I no longer call tslint from NPM scripts but it may be important for those people who do since NPM does not provide (and I believe that it never will) a cross platform glob support.

**Comment by Healforgreen** [comment]
I don't think it's a good idea to add options for files you want or don't want to lint in the `tslint.json` file.  It just doesn't seem like it fits there.  It also isn't consistent with how **jshint** is configured.  Yes, they are two different tools but they practically do the same thing. Since TS is a super set of JS there should be some consistency between the two.  I'd rather have a `.tslintignore` file (just like `.jshintignore`) to maintain consistency between JS and TS.  Keep it separate from `tsconfig.json` since linting is not compiling.  

+1 for the CLI option, it's a good start.

**Comment by JKillian** [comment]
Wanted to mention that #1006 merged and was in a release a month or so ago, so if you're using the latest version of the TSLint CLI, you'll have an `exclude` flag available for use.

**Comment by JoshuaKGoldberg** [comment]
For the sake of SEO, this keeps popping up for the "how to exclude .d.ts files from tslint" search.

```javascript
{
    exclude: "/.*\.d\.ts$/."
}
```

**Comment by ollwenjones** [comment]
+1

**Comment by mchambaud** [comment]
+1

**Comment by fyodorvi** [comment]
+1

**Comment by adidahiya** [comment]
what are people +1'ing here? an `--exclude` flag has been available in the CLI since v3.6 which allows you to exclude globs (files, paths, whatever).

**Comment by fyodorvi** [comment]
@adidahiya It doesn't help with WebStorm, for example. There's no option to specify flags.
![image](https://cloud.githubusercontent.com/assets/9676134/15730171/6c14734c-28be-11e6-85b2-4eadac7c8864.png)

I understand, that it should be a feature request for WebStorm developers, but it's seems that it's super easy to implement on ts-lint side. And that is done for other linting libraries by default. So why not make it so?

**Comment by adidahiya** [comment]
@fyodorvi cool, I updated the issue title and description. We'd likely want to resolve the path of the `.tslintignore` file the same way we resolve `tslint.json`.

Pros / cons of a new file vs. making it a new field in `tslint.json`?

**Comment by Karabur** [comment]
There is a workaround for ones who want to use tslint with WebStorm:
You can put empty (with '{}' inside) `tslint.json` file in your node_modules folder and enable 'Search for tslint.json` option in WS. That will completely reset all rules of tslint for node_modules folder, and it will be silent about vendor code (or you can add specific rules here if you want to)

**Comment by bill-mybiz** [comment]
I've been following this thread to be able to exclude files specifically using vscode (since before the name was changed to be .tslintignore specific). I wanted to let others know who are trying to integrate the exclude with vscode that there is a [separate issue](https://github.com/Microsoft/vscode-tslint/issues/23) and [recent PR](https://github.com/Microsoft/vscode-tslint/pull/40) for vscode-specific tslint excludes.

**Comment by slayerfat** [comment]
@adidahiya a field with an array of dirs or files inside tslint.json should suffice, i would rather not worry about another file to control this business when we already have a config file defined in a project.

@Karabur your workaround doesn't work for me, i even tried to add an empty rules object, am i missing something?

**Comment by NickolasAcosta** [comment]
I'd rather not have another file personally.  One is enough.

**Comment by basarat** [comment]
An `exclude` option in `tslint.json` would be nice (only one file to use that way) and that is the name of the `glob` ignore section of `tsconfig.json` as well (yes tsconfig.json include / exclude now supports globs in TypeScript latest)

Anyone know what eslint does as I am sure they would have hit this too?  :rose:

**Comment by JKillian** [comment]
Right now, `tslint.json` has no control over _what_ is linted, it only controls _how_ files are linted. In fact, depending on how you set things up, the TSLint CLI can actually lint different files by different rules all in the same run.

I'm not against an `exclude` option in `tslint.json`, but it might end up being a slight paradigm shift and would have to be thought about carefully.

P.S. Always happy to hear your feedback @basarat, was just reading your TS gitbook today 😉

**Comment by basarat** [comment]
I've just released a new version of [alm with tslint intergrated](http://alm.tools/) with version `0.3.0` :heart:

<img width="703" alt="linter" src="https://cloud.githubusercontent.com/assets/874898/17267065/cccd14e2-5644-11e6-97b0-b6c61b0fb5a3.png">

After having it sit in my brain here is the summary of the design I've gone with for IDE integration and seems to work out beautifully: 

* all files in the current `tsconfig.json` are linted except for `.d.ts` files
* `tslint.json` should be determined using a `findup` from `tsconfig.json`. 

I've documented this here as well : https://basarat.gitbooks.io/alm/content/features/lint.html Feel free to use the ideas / source for any other IDE integrations :rose:

**Comment by JKillian** [comment]
@basarat I think those choices sound perfect! I haven't tried alm yet, but I'll have to give it a try sometime soon 💻

**Comment by itaysabato** [comment]
+1

**Comment by oka-garaio** [comment]
+1

**Comment by ataraxus** [comment]
+1

**Comment by adidahiya** [comment]
Please stop posting +1 comments and use the built-in Github upvoting feature. I'm going to delete your +1 comments now.

**Comment by maksimluzik** [comment]
Two years have passed, and this is still not implemented? How come, this is one of the most trivial and important features IMO.

**Comment by abenhamdine** [comment]
> Three years have passed, and this is still not implemented? How come, this is one of the most trivial and important features IMO.

It's OSS after all : if you want it, submit a PR.

**Comment by monolithed** [comment]
The current problem for me is:
```
--exclude {node_modules,.git,cache}
```

The `{}` operator does not works and I don't see any way how to exclude different directories...

**Comment by JKillian** [comment]
@monolithed: 

```
--exclude node_modules --exclude .git --exclude cache
```

should do what you want

**Comment by mmkal** [comment]
> > Three years have passed, and this is still not implemented? How come, this is one of the most trivial and important features IMO.

> It's OSS after all : if you want it, submit a PR.

I've submitted a PR: https://github.com/palantir/tslint/issues/2409

@abenhamdine any ideas whom I should ask to review?

**Comment by machineghost** [comment]
Bumping a 3+ year old issue wanted by literally hundreds of people, with a working PR submitted, but no action for half a year.

**Comment by adidahiya** [comment]
Thanks for the bump. Note that the linked PR, #2409, is for a new config option in `tslint.json`, which is not quite the same as the original issue (`.tslintignore` file). It's a step in the right direction though; I've left some more CR comments.

**Comment by Domvel** [comment]
+1
Options:
- File `.tslintignore` to the folder to ignore. 
- File tslint.json (global file) new property `exclude` with a list of pathes / folders.
This would be nice.

**Comment by saitonakamura** [comment]
By the way, feature was released in `tslint@5.8.0`

**Comment by Domvel** [comment]
Right. Example: 
Create a folder in your project with name `.vscode` and a file inside `settings.json`. 
```json
{
  "tslint.exclude": "**/app/models/**"
}
```

**Comment by paranoidjk** [comment]
+1

**Comment by didilinkin** [comment]
+1 
need `.tslintignore`

**Comment by willclarktech** [comment]
Hi @JoshuaKGoldberg, that's a great start. I think what a lot of people on this thread were hoping for from a `.tslintignore` file besides functionality was that it would conform to the standard already set by git, npm, eslint etc. People like this standard, feel very comfortable with it, and kind of expect it - for example I found this issue when I created a `.tslintignore` file assuming it would work, and was surprised (rightly or wrongly) when it didn't.

**Comment by JoshuaKGoldberg** [comment]
Hmm. I'm dubious of having two standards for how this should work. If you have both `.tslintignore` and `linterOptions.exclude` in a directory, how does the behavior work?

**Comment by JoshuaKGoldberg** [comment]
Given #4534 and the `linterOptions.exclude`  workaround, this seems out of scope for TSLint now. The recommended approach is to migrate to [typescript-eslint](https://typescript-eslint.io); if you need to exclude some TypeScript files from TSLint linting, use `linterOptions.exclude`. Closing this issue for housekeeping.

If there is some blocker in `linterOptions.exclude` that's making it more difficult to migrate over, please do file a new issue to report the bug!

**Comment by JoshuaKGoldberg** [comment]
🤖 Beep boop! 👉  **TSLint is deprecated** 👈  and you should **switch to [typescript-eslint](https://typescript-eslint.io)**! 🤖 

🔒 This issue is being locked to prevent further unnecessary discussions. Thank you! 👋
